### DISCUSSION TITLE
Audio issues when running in systemd unit/service...

### DISCUSSION THREAD

**Discussion by goodlucknow** [ehough/docker-kodi#42] [Q&A] [answered]
Have revisited docker-kodi after some time and have managed to get it all working perfectly!

`x11docker --xorg --user=kodi --pulseaudio --gpu --wm=none --home=/home/kodi/ erichough/kodi` 
is the command that works for me.

However, when I try to run it from a systemd unit, I get no audio. Unit file is as follows:

```
[Unit]
Description=Dockerized Kodi
Requires=docker.service
After=network-online.target docker.service

[Service]
ExecStart=/usr/bin/x11docker --xorg --user=kodi --pulseaudio --gpu --wm=none --home=/home/kodi/ erichough/kodi
Restart=always
KillMode=process

[Install]
WantedBy=multi-user.target
```

From the container logs I can see some errors relating to jack server. Any ideas how I might fix, or what changes when running as a service vs running from command line?

```
---> starting Kodi with command: kodi-standalone
Connection failure: Connection refused
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started

```

Thanks in advance for any pointers and help!

**Comment by mviereck** [comment, marked as answer]
Normally the pulseaudio server is started after a user login and does not run system wide.
If you run x11docker from cli, there is normally a pulseaudio server already running.

But when systemd starts the services there is no pulseaudio that x11docker could connect to. 
It might help to use option `--alsa` instead of `--pulseaudio`.

About jack and how it interferes with pulseaudio or ALSA I cannot tell anything.

**Reply by goodlucknow** [reply]
Interesting, thank you for the insight.
I'll try out alsa and see, although the reason I'm using pulseaudio in the first place was issues with alsa and getting speaker channel mapping to work on my system!

**Reply by goodlucknow** [reply]
Your solution is correct, thank you. Alsa works within the systemd service.
Which leaves me with my original channel mapping issue, but that belongs on a different forum.
I don't suppose you would have any idea how to get pulseaudio working in this way? I'm off to google some more...
Thanks again for the tip.

**Reply by mviereck** [reply]
It is possible to set up a system wide pulseaudio server: https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/User/SystemWide/
This setup is rather discouraged: https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/User/WhatIsWrongWithSystemWide/

x11docker should be able to access a system wide pulseaudio server with `--pulseaudio=tcp` if its service is started before the x11docker unit.
Related, but without systemd: https://github.com/mviereck/x11docker/issues/266

**Reply by goodlucknow** [reply]
Given how discouraged system wide pulseaudio is, I am reluctant to use it as a solution... But I will not rule it out if there are no better options.
Would you think it is possible to connect to a user instance of pulseaudio?
With my current non-dockerized setup, I followed the example of [Benjamin Brummer](https://benjamin-brummer.de/2018/06/13/kodi-auf-ubuntu-server-18-04-lts-bionic-beaver/).

There is a kodi user configured without login and password, and the systemd service is as follows:

```
[Unit]
Description=Starts instance of Kodi using xinit
After=systemd-user-sessions.service network-online.target sound.target
Requires=network-online.target
Conflicts=getty@tty1.service

[Service]
User=kodi
Group=kodi
PAMName=login
TTYPath=/dev/tty1
ExecStart=/usr/bin/xinit /usr/bin/kodi-standalone -- :0 -nolisten tcp vt1
Restart=on-abort
StandardInput=tty

[Install]
WantedBy=multi-user.target
```

Is there any way of running x11docker similarly?

Apologies if I am asking at the wrong questions, I am quite out of my depth trying to achieve this setup.
When trying to configure the service in this way myself, x11docker seems to require a sudo password. I'm not sure whether to continue to troubleshoot that or if it is destined to fail anyhow...

**Reply by goodlucknow** [reply]
I am going to delve further into the x11docker wiki pages this evening to see if I can improve my understanding.

**Comment by mviereck** [comment]
>  Would you think it is possible to connect to a user instance of pulseaudio?

If the x11docker service is a system service and that user logs in later, I don't see a straightforward way. Some hacky solution might be possible, though.

After all, why do you want to run x11docker+kodi as a systemd service at all?

There are two sorts of systemd services: the system services (the common use case) and user services.
It might make sense to run as a user service with `systemd --user kodi.service` after the user has logged in. That should allow access to the users pulseaudio server immediately.

> I am quite out of my depth trying to achieve this setup.

This setup is at the border of my knowledge, too. :-)

> When trying to configure the service in this way myself, x11docker seems to require a sudo password.

docker needs root privileges to run, or the user must be in group `docker`. The service runs as user `kodi` who is not root and likely not in group `docker`. You could add user `kodi` to group `docker` (though, not recommended in general).
(rootless docker is possible, too, but not supported in x11docker for now. rootless setups supported by x11docker yet are `podman` and `nerdctl`.)

**Reply by mviereck** [reply]
[double post removed]

**Reply by goodlucknow** [reply]
Thank you for all the great info.

**Reply by mviereck** [reply]
Could you tell me why you want to run as a systemd service? It is a serious question, and if I understand the reason, I might have further ideas.

**Reply by goodlucknow** [reply]
Oh sorry, really the only reason is to have it start with the system at boot, and as I was moving from the setup as detailed above with kodi as a systemd service, and also it is the provided method in the examples for docker-kodi.
I am running a headless ubuntu server, and kodi on a connected display. Recently I have experienced some kodi crashes that froze the system, and thus all my docker containers & services. And so I would like to have kodi running inside a docker, to isolate it & keep the system alive in the event of a crash...
Does this make sense?

**Reply by goodlucknow** [reply]
adding `kodi` to the `docker` group indeed solves that problem, the service now runs, and audio works correctly! but I am aware this may be a security issue. I think for now it may be preferable to running pulseaudio systemwide.

**Comment by mviereck** [comment]
> adding kodi to the docker group indeed solves that problem, the service now runs, and audio works correctly!

Great!

> but I am aware this may be a security issue.

For rootless containers I recommend `podman` instead of docker. You can run it with x11docker option `--backend=podman`. That avoids the need of a root password or group `docker`.

> I think for now it may be preferable to running pulseaudio systemwide.

The concerns of systemwide pulseaudio are a different thing than concerns about group docker.
Maybe it is possible to run two user instances of pulseaudio server, one for user kodi and one for your regular user. I don't know if and how they would work together.

> Recently I have experienced some kodi crashes that froze the system, and thus all my docker containers & services. And so I would like to have kodi running inside a docker, to isolate it & keep the system alive in the event of a crash...

Hopefully that helps.

Edit: Arrgh, confused the answer boxes again.

**Reply by goodlucknow** [reply]
I will have a look into podman, it was not on my radar prior to this.
Just to be clarify - I would happy to give kodi exclusive use of pulseaudio, as with the display. I don't know if there is a better way of running this with that information in mind.
The only reason I need to run pulseaudio instead of alsa is the aforementioned problem with speaker remapping. This was an issue for me with Kodi installed directly on the system prior to docker and one I spent quite some time on with no solution.
Once again thanks for all your help and guidance.

**Reply by mviereck** [reply]
> Just to be clarify - I would happy to give kodi exclusive use of pulseaudio, as with the display. I don't know if there is a better way of running this with that information in mind.

Oh, ok, this should make things much easier.

Proposal:
Replace line:
```
ExecStart=/usr/bin/x11docker --xorg --user=kodi --pulseaudio --gpu --wm=none --home=/home/kodi/ erichough/kodi
```
with:
```
ExecStart=/usr/local/bin/kodicontainer
```
Create an executable script `/usr/local/bin/kodicontainer` with the content:
```
#! /bin/bash
pulseaudio -D # run pulseaudio daemon
/usr/bin/x11docker --xorg --user=kodi --pulseaudio --gpu --wm=none --home=/home/kodi/ erichough/kodi
```
If you decide to use podman, add `--backend=podman` and change the image name to `docker.io/erichough/kodi`.

**Reply by goodlucknow** [reply]
This results in some strange behaviour - on first execution(and after subsequent system reboots) the pulseaudio daemon fails to start, and I see the same issues as the original post above.
If I then restart the service, pulseaudio works correctly...

**Reply by mviereck** [reply]
Hmm.
Is pulseaudio started elsewhere, too?
Maybe it works better if there is a pulseaudio service (also with `USER kodi`, and without option `-D`) that is started before the kodi service.
The kodi service would need an entry like `After=pulseaudio-kodi.service`

**Reply by mviereck** [reply]
Just a thought:
pulseaudio has some sort of autostart feature.
Maybe the pulseaudio daemon is started automatically as soon as x11docker calls it. So there is no need to care for the pulseaudio daemon at all.

**Reply by goodlucknow** [reply]
I had a similar thought - removed that line from the script and it works.
This doesn't change the need to have the kodi user in the docker group though does it? In which case is there any benefit to abstracting the x11docker startup into a seperate script?

**Comment by mviereck** [comment]
> I had a similar thought - removed that line from the script and it works.

Ok, good.

> This doesn't change the need to have the kodi user in the docker group though does it? 

No. You could remove kodi from group `docker` if you use `podman` instead.

> In which case is there any benefit to abstracting the x11docker startup into a seperate script?

This was just for the purpose to start the pulseaudio damon, too. Without this there is no benefit.

**Reply by goodlucknow** [reply]
Ah, I understand. Thanks for your help. I will investigate podman in due course.
